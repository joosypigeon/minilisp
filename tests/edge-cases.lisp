;; Edge case: empty list
(print (if (equal? NIL NIL) "pass: empty list equals NIL" "fail: empty list equals NIL"))

;; Edge case: empty lambda parameters
(define empty-fn (lambda () 42))
(print (if (equal? (empty-fn) 42) "pass: empty lambda args" "fail: empty lambda args"))



;; Edge case: list with only one item
(define singleton (list 42))
(print (if (equal? (car singleton) 42) "pass: singleton list car" "fail: singleton list car"))
(print (if (equal? (cdr singleton) NIL) "pass: singleton list cdr" "fail: singleton list cdr"))

;; Edge case: shadowing symbol in let
(define y 5)
(let ((y 999)) (print (if (equal? y 999) "pass: let shadowing" "fail: let shadowing")))
(print (if (equal? y 5) "pass: let didn't overwrite" "fail: let didn't overwrite"))

;; Edge case: redefining a function name
(define dup (lambda (x) x))
(define dup (lambda (x) (+ x x)))
(print (if (equal? (dup 3) 6) "pass: function redefinition" "fail: function redefinition"))

;; Edge case: applying a non-lambda symbol
(print
  (if (equal?
        (begin
          (define not-a-fn 42)
          (if (equal? not-a-fn not-a-fn) "pass: non-callable" "fail: non-callable"))
        "pass: non-callable")
      "pass: handled non-callable"
      "fail: handled non-callable"))

;; Edge case: quoting complex structure
(define complex-quote '(a (b (c d)) e))
(print (if (equal? (car complex-quote) 'a) "pass: quote car" "fail: quote car"))
(print (if (equal? (car (cdr complex-quote)) '(b (c d))) "pass: quote cdr" "fail: quote cdr"))
