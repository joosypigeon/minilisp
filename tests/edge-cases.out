"pass: empty list equals NIL"
"pass: empty lambda args"
"pass: singleton list car"
"pass: singleton list cdr"
"pass: let shadowing"
"pass: let didn't overwrite"
"pass: function redefinition"
"pass: handled non-callable"
"pass: quote car"
"pass: quote cdr"
